capstone/src/test/java/com/google/sps/servlets/profile/ProfileServletTest.java

```diff
@@ -0,0 +1,249 @@
+import static org.mockito.Mockito.when;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.sps.data.UserProfile;
+import java.io.*;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class ProfileServletTest {
+
+  private LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+          .setEnvIsAdmin(true)
+          .setEnvIsLoggedIn(true);
+
+  @Mock private HttpServletRequest request;
+
+  @Mock private HttpServletResponse response;
+
+  @Mock private UserService userService;
+
+  @Mock private DatastoreService datastore;
+
+  private static final String NAME = "John Doe";
+  private static final String LOCATION = "Mountain View, CA";
+  private static final String BIO = "This is my bio.";
+  private static final String USERID = "12345";
+  private static final String INVALID_USERID = null;
+  private static final String EMAIL = "abc@gmail.com";
+  private static final String AUTHDOMAIN = "gmail.com";
+  private static final String PATHINFO = "profile/12345";
+  private static final String INVALID_PATHINFO = "profile";
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    helper.tearDown();
+  }
+
+  /*
+   *  Test doGet() for when user enters an invalid URL param. It will return an error.
+   **/
+  @Test
+  public void invalidUrlParamReturnError() throws ServletException, IOException {
+    when(request.getPathInfo()).thenReturn(INVALID_PATHINFO);
+
+    ProfileServlet userServlet = new ProfileServlet(userService, datastore);
+    userServlet.doGet(request, response);
+
+    // verify if a sendError() was performed with the expected values.
+    verify(response)
+        .sendError(
+            HttpServletResponse.SC_NOT_FOUND,
+            "The profile you were looking for was not found in our records!");
```

Comment:
```suggestion
            Mockito.anyString());
```
Eashan told me to do this on my PR. The advantage is that if someone changes the error message the test still passes. (Which it should, I think).

Reply:
I agree! Thanks :)